### PR TITLE
Emags have an increasing chance to explode when used by crew

### PR DESCRIPTION
**Pull request by Rex9001**
## About The Pull Request
PR requested by ZephyrTFA

When any non-traitor, spy, or nuclear operative (actually checked via traits and not datums, but gameplay wise this is how it functions) uses an emag it now has a 25% chance to explode, this increases by a +25% for each subsequent use, so (25,50,75,100) for (1,2,3,4) uses respectively. By explosion its very small, just a one range light explosion. When its going to explode the emag beeps and you are given 2 seconds to toss it away before the battery on it explodes! The explosion is triggered after the emag effect though so you will always get at least one emag use from it.
## Why It's Good For The Game
The emag is intended to be an antagonist item and shouldnt allow the crew to easily gain permanent buffs from it like being able to purchase autorifles. This also lets us lock more powerful effects behind emagging as this will decrease the chance of crew being able to unlock that powerful effect (assuming it requires multiple emags). This also adds an interesting risk/reward element to the emag and is a lot more open ended than just "crew cant use it hurr", instead the crew are given a choice, use the emag on a particular machine at the risk of being unable to use it for another, maybe the cargo console needs emagging because sec wants guns but the chemists also want their chem machines emagged for miners salve, in these situations the crew will need to actually strategize the use of the item instead of just obtaining what is essentially an upgrade card for a lot of devices.
## Changelog
:cl:
balance: The emag now has an increasing chance to explode when used by crew
/:cl:

### CODE REVIEW

**Comment by Pickle-Coding** [comment]
It's funnier if there isn't a delay to the explosion.

**Comment by Rex9001** [comment]
> It's funnier if there isn't a delay to the explosion.

Up to personal preference, but I always find like ominous beeping pretty funny personally.

File: code/game/objects/items/emags.dm [hunk 1]
```diff
@@ -73,6 +88,22 @@
...
+/obj/item/card/emag/proc/crew_action(mob/user)
+	if(!HAS_TRAIT(user, TRAIT_EMAG_USER))
+		crew_uses += 1
+	if(!prob(crew_uses * 25))
+		return
+	to_chat(user, span_warning("You hear a buzzing from [src] 'Non-authorised personel detected. Terminating card'."))
+	user.visible_message(span_boldwarning("[src] beeps ominously!"))
+	playsound(loc, 'sound/items/timer.ogg', 35, vary = FALSE)
+	// Think fast chucklenuts
+	stoplag(2 SECONDS)
```
Comment (ZephyrTFA) [inline code/game/objects/items/emags.dm:101]: use a timer not a sleep

File: code/game/objects/items/emags.dm [hunk 2]
```diff
@@ -73,6 +88,22 @@
 			return FALSE
 	return TRUE
 
+/// What happens if a crewmember uses this card
+/obj/item/card/emag/proc/crew_action(mob/user)
+	if(!HAS_TRAIT(user, TRAIT_EMAG_USER))
+		crew_uses += 1
+	if(!prob(crew_uses * 25))
+		return
+	to_chat(user, span_warning("You hear a buzzing from [src] 'Non-authorised personel detected. Terminating card'."))
```
Comment (ZephyrTFA) [inline code/game/objects/items/emags.dm:97]: needs to add a flag to prevent the emag from just being used until it actually goes boom

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]

File: code/game/objects/items/emags.dm [hunk 3]
```diff
@@ -73,6 +88,22 @@
...
+	user.visible_message(span_boldwarning("[src] beeps ominously!"))
```
Comment (MrMelbert) [inline code/game/objects/items/emags.dm:98]: \>"beeps"
\>`visible_message` and not `audible_message`

File: code/game/objects/items/emags.dm [hunk 4]
```diff
@@ -73,6 +88,22 @@
...
+	playsound(loc, 'sound/items/timer.ogg', 35, vary = FALSE)
+	// Think fast chucklenuts
+	stoplag(2 SECONDS)
+	user.visible_message(span_boldwarning("[src]s battery ruptures!"))
```
Comment (MrMelbert) [inline code/game/objects/items/emags.dm:102]:
```suggestion
	user.visible_message(span_boldwarning("[src]'s battery ruptures!"))
```

File: code/game/objects/items/emags.dm (hunk 1 above)
Comment (MrMelbert) [inline code/game/objects/items/emags.dm:101]: Use a timer

File: code/game/objects/items/emags.dm [hunk 5]
```diff
@@ -73,6 +88,22 @@
...
 
+/// What happens if a crewmember uses this card
+/obj/item/card/emag/proc/crew_action(mob/user)
+	if(!HAS_TRAIT(user, TRAIT_EMAG_USER))
+		crew_uses += 1
+	if(!prob(crew_uses * 25))
+		return
+	to_chat(user, span_warning("You hear a buzzing from [src] 'Non-authorised personel detected. Terminating card'."))
+	user.visible_message(span_boldwarning("[src] beeps ominously!"))
+	playsound(loc, 'sound/items/timer.ogg', 35, vary = FALSE)
```
Comment (MrMelbert) [inline code/game/objects/items/emags.dm:99]:
```suggestion
	playsound(src, 'sound/items/timer.ogg', 35, vary = FALSE)
```

File: code/game/objects/items/emags.dm (hunk 1 above)
Comment (Rex9001) [reply code/game/objects/items/emags.dm:101]: sure

**Comment by Formerxboxgamer** [comment]
Antag test and a really lame way to limit the sandbox. We shouldn’t be allergic to crew using antag items if an antag loses them.

**Comment by Metekillot** [comment]
Agree with Formerxbox.

**Comment by Vect0r2** [comment]
If we have to snowflake the Emag effects because they are so powerful, maybe we are giving too much power to a 4TC item.

**Comment by DaydreamIQ** [comment]
Antag testing aside I think this is hilarious. I suppose you could add some kind of way for antags to make the emag self destruct if people are THAT worried about antag checking

**Comment by Rex9001** [comment]
Im honestly not very concerned about antag testing for this item, I was considering a triggerable self destruct effect and might do it in a follow up if it proves to be an issue. Tbh getting found with an emag usually means its already over for you.

**Comment by ItzRiumz** [comment]
A major problem is that security could use this to test every suspected traitor/spy in the round. So sure, the guy who originally got caught is screwed, but the other antags in the round are also screwed if security can force them to emag things to test them.

**Comment by SmArtKar** [comment]
This PR was made after a discussion in #coding-general, during which it was decided that if emag gets restricted/nerfed for crew usage we could go more wild with it's effects (nullcrates?). Blowing up after 3-4 uses is a good way to do that. This isn't a nerf, this is a prelude to potential buffs/features.

**Comment by Jacquerel** [comment]
it's not a very effective antag test if as soon as you test someone who isn't an antag it explodes (and you get bwoinked for delimbing a random guy)

**Comment by alien3301** [comment]
This doesn't make much sense IC to me. Isn't the emag's design incredibly simple? How could you fuck up that badly when using it.

**Comment by Jacquerel** [comment]
> This doesn't make much sense IC to me. Isn't the emag's design incredibly simple? How could you fuck up that badly when using it.

I mean from looking at the sprite it's surprising it doesn't explode more often, that thing is jury-rigged to hell

**Comment by Formerxboxgamer** [comment]
> This PR was made after a discussion in #coding-general, during which it was decided that if emag gets restricted/nerfed for crew usage we could go more wild with it's effects (nullcrates?). Blowing up after 3-4 uses is a good way to do that. This isn't a nerf, this is a prelude to potential buffs/features.

why not add a new uplink item rather than restricting the sandbox? What this will do is force people to only emag stuff like the cargo console or chem dispenser, and not bother emagging funny stuff like sec hailers. I don't want to see people be forced to only emag the optimal gamer stuff instead of walking around the station like a kid in a candy shop seeing what items have emag interactions.

**Comment by ATH1909** [comment]
> This PR was made after a discussion in #coding-general, during which it was decided that if emag gets restricted/nerfed for crew usage we could go more wild with it's effects (nullcrates?). Blowing up after 3-4 uses is a good way to do that. This isn't a nerf, this is a prelude to potential buffs/features.

IMO, those wilder effects should be their own traitor items, not rolled into the emag (which, as someone mentioned earlier, is 4 TC and already an incredibly useful item). And you still have to design around sec getting to use at least one of those wilder effects- is  sec having "only" 1-4 traitor mechs really that much better than sec having as many as they have the mats to print?

Speaking of mats, I kind of think that external resources are a better way to limit these wilder effects than non-antag emag charges. Like, science can also print cracked stuff, but they're held back by research requirements, mat requirements, effort requirements, and anomaly core requirements, so you don't see sec running around with an army of Phazons every shift. Isn't "hey, making this costs synthflesh, an anomaly core, some telecrystals, and/or these esoteric components" a better way to limit the non-antag usage of wackier emag effects?

**Comment by Higgin** [comment]
this isn't it

you lose an emag, the crew gets an emag. 

don't lose an emag.

don't snowflake it.

**Comment by Jacquerel** [comment]
> why not add a new uplink item rather than restricting the sandbox? What this will do is force people to only emag stuff like the cargo console or chem dispenser, and not bother emagging funny stuff like sec hailers. I don't want to see people be forced to only emag the optimal gamer stuff instead of walking around the station like a kid in a candy shop seeing what items have emag interactions.

The intention of this change is to limit non-antagonist use of emags so saying "but this will limit non-antagonist use of emags!" is not likely to be a persuasive argument
this doesn't do anything for antagonist use of emags because they're not effected by this change

**Comment by Metekillot** [comment]
I think a better way to discourage crew use of the emag (because who gives a shit about losing an arm if I get sweet traitor perks baybee) is to make the emag give you a permanent brain trauma due to syndicate counter hacking your use of their Super Sekrit Syndicate Software Algorithms.

**Comment by Formerxboxgamer** [comment]
> > why not add a new uplink item rather than restricting the sandbox? What this will do is force people to only emag stuff like the cargo console or chem dispenser, and not bother emagging funny stuff like sec hailers. I don't want to see people be forced to only emag the optimal gamer stuff instead of walking around the station like a kid in a candy shop seeing what items have emag interactions.
> 
> The intention of this change is to limit non-antagonist use of emags so saying "but this will limit non-antagonist use of emags!" is not likely to be a persuasive argument this doesn't do anything for antagonist use of emags because they're not effected by this change

Cool, but frankly the idea is principally flawed and I'm pointing that out. Saying that my criticism isnt valid because I dont like the core of the PR is weird,  you're acting like limiting crew use of an emag being good is a truism.

**Comment by Jacquerel** [comment]
> Cool, but frankly the idea is principally flawed and I'm pointing that out. Saying that my criticism isnt valid because I dont like the core of the PR is weird, you're acting like limiting crew use of an emag being good is a truism.

What I am saying is that it's probably not going to be an effective argument to get anything you want from the person who made the PR; you are free to share it but it's probably a waste of your time.

**Comment by Metekillot** [comment]
I'm against having foolproof antag tests for traitors, so I'll reiterate support for making this a stealthier penalty rather than very obviously exploding.

**Comment by breadcrombs** [comment]
when are you going to make it so that makarovs and revolvers instantly misfire on crew, randomly becoming clumsy when holding a esword, and stims or syndie modsuits instantly dust non antags upon use

**Comment by moocowswag** [comment]
I thought the previous narrative for emag was it was over bloated with so many different effects now we made a 180 degree turn and want to nerf crew's ability to use it so we can bloat the item even more?

Giving emag/hacked effects even more powerful uses (granted there are already plenty) just further solidifies it as a must buy for every traitor. Go down this road and itl see a higher tc/rep requirement.

**Comment by Metekillot** [comment]
So?

**Comment by moocowswag** [comment]
> So?

You cant balance emag effects to be emag only ,-, malf ai and saboteur borgs (and i think ninja?) also have access to them

**Comment by moocowswag** [comment]
If you make it emag only suddenly the consistency of all of these antagonists unlocking a hacked state of any machine is shattered but yes you can code whatever you want.

**Comment by Metekillot** [comment]
Yes, it may break an established pattern, or it may not. Or it may be factored in that an AI or borg with an emag can hack it. The possibilities are limitless and we're just talking in circles.

EIther way, I don't think it should explode. I think it should be some stealthy, delayed penalty, that's quite profound instead of a momentary inconvenience. A light explosion on your tile will delimb you, which can be remedied by a competent medbay in 3 minutes, or an incompetent medbay in 10-35 minutes.

**Comment by Eronymun** [comment]
Normally I'm open for big nerfs if I think they'll have a net positive impact, but I'm not really in favor of this.

I think just increasing the cost of the emag to 8 TC would be a lot healthier. Limiting the potential for cascading events due to someone having an emag when they shouldn't feels like it would hurt some of the sandbox potential in rounds.

File: code/game/objects/items/emags.dm [hunk 6]
```diff
@@ -14,8 +14,16 @@
...
 	slot_flags = ITEM_SLOT_ID
 	worn_icon_state = "emag"
-	var/prox_check = TRUE //If the emag requires you to be in range
-	var/type_blacklist //List of types that require a specialized emag
+	/// If the emag requires you to be in range
+	var/prox_check = TRUE
+	/// List of types that require a specialized emag
+	var/type_blacklist
+	/// If the emag can be used by crewmembers
+	var/useable_about_crew = FALSE
```
Comment (necromanceranne) [inline code/game/objects/items/emags.dm:22]: The spelling mistake aside, the vars not terribly accurate. You can use it as crew, but it is dangerous to do so.

Probably better to call it ``has_theft_countermeasure`` or something.

File: code/game/objects/items/emags.dm [hunk 7]
```diff
@@ -73,6 +94,25 @@
 			return FALSE
 	return TRUE
 
+/// What happens if a crewmember uses this card
+/obj/item/card/emag/proc/crew_action(mob/user)
+	if(!HAS_TRAIT(user, TRAIT_EMAG_USER))
+		crew_uses += 1
+	if(!prob(crew_uses * 25))
```
Comment (necromanceranne) [inline code/game/objects/items/emags.dm:101]: This should be before you increment the emag. This isn't so much as a means to still use it as crew at least one time, but as a way to create uncertainty about the emag user.

File: code/game/objects/items/emags.dm [hunk 8]
```diff
@@ -73,6 +94,25 @@
...
+	user.audible_message(span_boldwarning("[src] beeps ominously!"))
+	playsound(src, 'sound/items/timer.ogg', 35, vary = FALSE)
+	// Think fast chucklenuts
+	is_exploding = TRUE
+	addtimer(CALLBACK(src, PROC_REF(detonate), user), 2 SECONDS)
+
+/// What happens when this card is exploded! Used if you want to add any unique explosion effects to the card.
+/obj/item/card/emag/proc/detonate(mob/user)
+	user.visible_message(span_boldwarning("[src]'s battery ruptures!"))
+	explosion(src, light_impact_range = 1, explosion_cause = src)
```
Comment (necromanceranne) [inline code/game/objects/items/emags.dm:113]: If you really, REALLY, want to dissuade people using this as crew AND as an antag checker, and don't want to fuck with the probabilities, give this a dev range so it gibs the user. If sec start using this to antag check and get a normal crew member gibbed as a result, they're going to be in massive shit. Sure, it is a feels bad for the sap who got the emag, but it also gives our failed antagonist a final comeuppance against the crew.

**Comment by Holinka4ever** [comment]
I am fully understand that all traitor tools is intended to be used by agents only. But if the agent loose his gear due of death from being non efficient as antag, why we should like, give them positive points that their items will not be used by crew? What if someone on github due of these changes, gonna slap these functions to all agent loot? Like desword can cut your limb, noslips slips you, thermals blinds etc and etc, cause you are not valid yeah.
If agent loosing it, thats mean agent on blame that give powerfull item to crew, thats a point.

**Comment by JupiterJaeden** [comment]
Limits the sandbox potential of an item for no reason. Also introduces an antag testing method. The supposed benefit of this change does not outweigh the downsides. Emags are not obtained by crew often enough to be a serious balance concern, and the crew-side benefits that can be gained from them are 1) not strong enough to be overpowered, and 2) also still abusable by antags- any antag can use that same emagged cargo console to get those autorifles, or that emagged chem dispenser to make poisons.

**Comment by Blacklist897** [comment]
this is just a traitor tester and a snowflake change, sometimes it seems that we all forget that part of a space station round is randomness, a traitor losing a emag and cargo ordering guns with it is cool

**Comment by Mothblocks** [comment]
I think this is absurdly lame, I don't think I want too many of these bespoke antag only interactions in general but my primary memories relating to non-antag emag have been fine. I think if there are things we deem too powerful to emag then there's more general routes we can take, like nerfing the stuff that we find too stupid, or giving the emag a cooldown for extended uses, or whatever, but this change does not smell great to me at all

**Pull request closed without merging**